inspirehep/modules/workflows/utils.py

```diff
@@ -141,3 +145,54 @@ def _try_to_log(logfn, *args, **kwargs):
         return res
 
     return _decorator
+
+
+@with_debug_logging
+def get_pdf_in_workflow(obj):
+    """Return the fullpath to the PDF attached to a workflow object
+
+    Return the PDF attached to a workflow object. If no PDF is already attached
+    and the workflow object is related to an arXiv record, fetches on the fly
+    the PDF from arXiv and attaches it.
+    """
+
+    def _fetch_from_submitter_url(submission_pdf):
+        filename = secure_filename("fulltext.pdf")
+        pdf = download_file_to_workflow(
+            workflow=obj,
+            name=filename,
+            url=submission_pdf
+        )
+        if pdf:
+            pdf['doctype'] = 'main'
+            obj.log.info("PDF provided by user from {0}".format(submission_pdf))
+            return obj.files[filename].file.uri
+        else:
+            obj.log.info("Cannot fetch PDF provided by user from {0}".format(submission_pdf))
+
+    for filename in obj.files.keys:
+        if filename.endswith('.pdf'):
+            # There's a PDF already attached
+            return obj.files[filename].file.uri
+
+    submission_pdf = obj.extra_data.get('submission_pdf')
+    if submission_pdf:
+        return _fetch_from_submitter_url(submission_pdf)
+    obj.log.info("No PDF available")
+
+
+def download_file_to_workflow(workflow, name, url):
+    """Download a file to a specified workflow.
+
+    The ``workflow.files`` property is actually a method, which returns a
+    ``WorkflowFilesIterator``. This class inherits a custom ``__setitem__``
+    method from its parent, ``FilesIterator``, which ends up calling ``save``
+    on an ``invenio_files_rest.storage.pyfs.PyFSFileStorage`` instance
+    through ``ObjectVersion`` and ``FileObject``. This method consumes the
+    stream passed to it and saves in its place a ``FileObject`` with the
+    details of the downloaded file.
+    """
+    with closing(requests.get(url=url, stream=True)) as req:
+        if req.status_code == 200:
+            workflow.files[name] = req.raw
```

Comment:
@david-caro, @jacquerie here is the place where we risk to be overloaded by a too big requests. I don't see out of the box how we can inject here a check to stop reading the request after a configurable amount of bytes.
Any suggestion?